### PR TITLE
feat(code-block): client-side highlighting

### LINKED ISSUE #1091: [feat]: `<rh-code-block>` extensions
Today at Office hours we took the decision to release a very basic version of code-block with bellsprout.

features we'll push off to bulbasaur:
- new variants of `<rh-button>`: icon button / FAB for copy
- syntax highlighting
- callout badges
- toolbar/container

We also have some usability questions we'd like to address:
- should copy button always be a FAB (i.e. floating over the codeblock, not slotted into the toolbar)
- is combining tabs with links/actions a usability problem for sighted keyboard users?

SO, basically the bellsprout version will be a gray box that formats a script of text

_Originally posted by @bennypowers in https://github.com/RedHat-UX/red-hat-design-system/issues/328#issuecomment-1607508981_

### PR DESCRIPTION
**Pull request by bennypowers**
Closes #1091 

## What I did

1. add `highlighting="client` attribute
2. add `dedent` boolean attribute

## Testing Instructions

1. [/elements/code-block/demo/client-side-highlighting/](https://deploy-preview-1837--red-hat-design-system.netlify.app/elements/code-block/demo/client-side-highlighting/)

### CODE REVIEW

File: elements/rh-code-block/rh-code-block.css [hunk 1]
```diff
@@ -279,3 +286,76 @@
...
+  --_variable-color: var(--rh-color-orange-60, #9e4a06);
+}
+
+.on.dark {
+  --_default-color: var(--rh-color-gray-20, #e0e0e0);
+  --_selected-text-background: var(--rh-color-gray-95, #151515);
+  --_comment-color: var(--rh-color-gray-50, #707070);
+  --_comment-block-color: var(--rh-color-gray-50, #707070);
+  --_doctype-color: var(--rh-color-gray-50, #707070);
+  --_cdata-color: var(--rh-color-gray-50, #707070);
```
Comment (adamjohnson) [inline elements/rh-code-block/rh-code-block.css:333]: This color does not pass WCAG color contrast guidelines on dark.

File: elements/rh-code-block/rh-code-block.ts [hunk 2]
```diff
@@ -9,12 +9,23 @@ import { SlotController } from '@patternfly/pfe-core/controllers/slot-controller
...
+ */
+function dedent(str: string) {
+  const stripped = str.replace(/^\n/, '');
+  const match = stripped.match(/^\s+/);
+  return match ? stripped.replace(new RegExp(`^${match[0]}`, 'gm'), '') : str;
+}
+
 interface CodeLineHeightsInfo {
   lines: string[];
   lineHeights: (number | undefined)[];
```
Comment (adamjohnson) [inline elements/rh-code-block/rh-code-block.ts:31]: It looks like lines 51-80 need some reformatting to appease the [Code accordion in the docs](https://deploy-preview-1837--red-hat-design-system.netlify.app/elements/code-block/code/#rh-code-block).

If you click on "Attributes" in the accordion, you can see `actions` listed as an attribute. 

code-block-attributes-actions-docs

File: elements/rh-code-block/rh-code-block.ts [hunk 3]
```diff
@@ -172,6 +205,7 @@ export class RhCodeBlock extends LitElement {
           </svg>
         </button>
       </div>
+
```
Comment (adamjohnson) [inline elements/rh-code-block/rh-code-block.ts:208]: The `<button>` element on line 196 needs the `aria-expanded` and `aria-controls` attributes to communicate state to AT. Follow the APG guidelines for [disclosure](https://www.w3.org/WAI/ARIA/apg/patterns/disclosure/) to see how to implement.

**Review by adamjohnson — CHANGES_REQUESTED** [review]
Some a11y and docs changes. Otherwise, nice job. 👍

File: elements/rh-code-block/rh-code-block.css (hunk 1 above)
Comment (bennypowers) [reply elements/rh-code-block/rh-code-block.css:333]: CC @marionnegp

File: elements/rh-code-block/rh-code-block.ts (hunk 2 above)
Comment (bennypowers) [reply elements/rh-code-block/rh-code-block.ts:31]: recco to either defer this to another issue or, if Dr. Mr. @coreyvickery  selects icons, make it moot

File: elements/rh-code-block/rh-code-block.ts (hunk 3 above)
Comment (bennypowers) [reply elements/rh-code-block/rh-code-block.ts:212]: PTAL TY

Comment (adamjohnson) [reply elements/rh-code-block/rh-code-block.ts:212]: LGTM TY

**Comment by adamjohnson** [comment]
Axe DevTools complains that `#content-lines` needs to be focusable so that users can scroll the content within the `<div>`.

I propose we add `tabindex="0"` to that div when there are scrollbars / the code block is not expanded.

```html
<div id="content-lines" tabindex="0">
  ...
</div>
```

File: elements/rh-code-block/rh-code-block.css (hunk 1 above)
Comment (marionnegp) [reply elements/rh-code-block/rh-code-block.css:333]: `--rh-color-gray-60, #4D4D4D` for all of those

**Review by marionnegp — CHANGES_REQUESTED** [review]
Is it possible to use [these colors](https://github.com/RedHat-UX/red-hat-design-system/issues/1091#issuecomment-1917881422) for the prism js properties? I mostly want to make color combos like the one below have better contrast:
Screenshot 2024-09-10 at 1 51 44 PM

**Comment by adamjohnson** [comment]
[6418df9](https://github.com/RedHat-UX/red-hat-design-system/pull/1837/commits/6418df9b076fb7df869af82be9a36d8d45b84f62) implements tabindex, but in reverse. 

Right now, the `tabindex` attribute shows when the code block is expanded. It should only be there when the code block is _not_ expanded.

File: elements/rh-code-block/rh-code-block.ts [hunk 4]
```diff
@@ -128,21 +160,26 @@ export class RhCodeBlock extends LitElement {
...
+                               actions,
+                               compact,
+                               expandable,
+                               fullHeight,
+                               resizable,
+                               truncated,
+                               wrap })}"
            @code-action="${this.#onCodeAction}">
-        <div id="content-lines">
+        <div id="content-lines" tabindex="${ifDefined((fullHeight || undefined) && 0)}">
```
Comment (bennypowers) [inline elements/rh-code-block/rh-code-block.ts:172]:
```suggestion
        <div id="content-lines" tabindex="${ifDefined((!fullHeight || undefined) && 0)}">
```

**Pull request merged**